engine/crates/engine-v2/src/sources/graphql/federation.rs

```diff
@@ -91,34 +77,214 @@ impl FederationEntityPreparedExecutor {
+                    inputs: vec![(&self.operation.entities_variable_name, representations)],
+                };
+
+                tracing::debug!(
+                    "Query {}\n{}\n{}",
+                    subgraph.name(),
+                    self.operation.query,
+                    serde_json::to_string_pretty(&variables).unwrap_or_default()
+                );
+                let json_body = serde_json::to_string(&serde_json::json!({
+                    "query": self.operation.query,
+                    "variables": variables
+                }))
+                .map_err(|err| format!("Failed to serialize query: {err}"))?;
+
+                execute_subgraph_request(
+                    ctx,
+                    span.clone(),
+                    subgraph.name(),
+                    move || FetchRequest {
+                        url: subgraph.url(),
+                        headers: ctx.subgraph_headers_with_rules(subgraph.header_rules()),
+                        json_body,
+                        subgraph_name: subgraph.name(),
+                        timeout: subgraph.timeout(),
                     },
+                    ingester,
                 )
-                .deserialize(&mut serde_json::Deserializer::from_slice(&bytes))?;
-                Ok((status, subgraph_response))
-            },
-        )
+                .await
+            }
+        }
         .instrument(span);
+
         Ok(fut)
     }
 }
+
+struct EntityIngester<'ctx, R: Runtime> {
+    ctx: ExecutionContext<'ctx, R>,
+    plan: PlanWalker<'ctx, (), ()>,
+    cache_entries: Option<Vec<CachedEntity>>,
+    subgraph_response: SubgraphResponse,
+    cache_ttl: Duration,
+}
+
+pub struct CachedEntity {
+    pub key: String,
+    pub data: Option<Vec<u8>>,
+}
+
+impl<'ctx, R> ResponseIngester for EntityIngester<'ctx, R>
+where
+    R: Runtime,
+{
+    async fn ingest(self, bytes: Bytes) -> Result<(GraphqlResponseStatus, SubgraphResponse), ExecutionError> {
+        let Self {
+            ctx,
+            plan,
+            cache_entries,
+            mut subgraph_response,
+            cache_ttl,
+        } = self;
+
+        let status = {
+            let response = subgraph_response.as_mut();
+            GraphqlResponseSeed::new(
+                EntitiesDataSeed {
+                    response: response.clone(),
+                    cache_entries: cache_entries.as_deref(),
+                    plan,
+                },
+                EntitiesErrorsSeed {
+                    response,
+                    response_keys: plan.response_keys(),
+                },
+            )
+            .deserialize(&mut serde_json::Deserializer::from_slice(&bytes))?
+        };
+
+        if let Some(cache_entries) = cache_entries.filter(|_| status.is_success()) {
+            update_cache(ctx, cache_ttl, bytes, cache_entries).await
+        }
+
+        Ok((status, subgraph_response))
+    }
+}
+
+async fn update_cache<R: Runtime>(
```

Comment:
this could be a bit simpler if you kept only the cache_entries that need to be updated, as you do for the representations